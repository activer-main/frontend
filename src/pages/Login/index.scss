@import '/src/styles/variables';
@import '/src/styles/global';

.login{
  @include container-main;

  display: flex;
  flex: 1 1 auto;
  flex-direction: column;
  gap: 1em;
  justify-content: center;
  max-width: 30em;

  &__header{
    text-align: center;
    
  }
  
  &__control{
    display: flex;
    gap: 1em;
    justify-content: center;
  }
}