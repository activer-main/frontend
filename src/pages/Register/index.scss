@import '/src/styles/variables';
@import '/src/styles/global';

.register{
  @include login-out;
}