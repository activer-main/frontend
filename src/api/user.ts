import { UserInfoType } from 'types/user';
import axios from 'axios';
import { LoginResponseType, RegisterResponseType } from 'types/response';
import { RegisterRequestType, LoginRequestType, UserUpdateRequestType } from 'types/request';
import { userToken } from 'store/auth/authSlice';

const IP = '220.132.244.41';
const PORT = '5044';

export const URL = `http://${IP}:${PORT}`;

const userResquest = axios.create(
  {
    baseURL: URL.concat('/api/user'),
  },
);

export const postRegist = (
  registBody : RegisterRequestType,
) => userResquest.post<RegisterResponseType>(
  '/auth/signup',
  registBody,
  {
    headers: {
      'Content-Type': 'application/json',
    },
  },
);

export const postLogin = (loginBody: LoginRequestType) => userResquest.post<LoginResponseType>(
  '/auth/signin',
  loginBody,
  {
    headers: { 'Content-Type': 'application/json' },
    withCredentials: false,
  },
);

// eslint-disable-next-line
export const putUserData = (newUserInfo : UserUpdateRequestType) => userResquest.put<UserInfoType>(
  '',
  {
    RealName: 'John Doe',
    NickName: 'johndoe123',
    Avatar: 'https://example.com/avatar.png',
    Gender: 'male',
    Birthday: '1980-01-01T00:00:00.000Z',
    Profession: 'Software Developer',
    Phone: '123-456-7890',
    County: 'Los Angeles',
    Area: 'Santa Monica',
    ActivityHistory: ['activity1', 'activity2', 'activity3'],
    TagHistory: ['tag1', 'tag2', 'tag3'],
  },
  {
    headers: {
      'Content-Type': 'multipart/form-data',
      Authorization: `Barear ${userToken}`,
    },
  },
);
