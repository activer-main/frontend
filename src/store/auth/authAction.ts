import { LoginResponseType, RegisterResponseType } from 'types/response';
import { postRegist } from 'api/user';
import { RegisterRequestType } from 'types/request';
import { createAsyncThunk } from '@reduxjs/toolkit';
import { UserInfoType } from '../../types/user';
import { LoginRequestType } from '../../types/request';
import { postLogin } from '../../api/user';

// Register
export const registerUser = createAsyncThunk<
RegisterResponseType,
RegisterRequestType
>(
  'auth/register',
  async (registBody: RegisterRequestType, { rejectWithValue }) => {
    try {
      const { data } = await postRegist(registBody);
      return data;
    } catch (error: any) {
    // return custom error message from backend if present
      if (error.response && error.response.data.message) {
        return rejectWithValue(error.response.data.message);
      }
      return rejectWithValue(error.message);
    }
  },
);

// Login
export const userLogin = createAsyncThunk<
LoginResponseType,
LoginRequestType
>(
  'auth/login',
  async (loginBody: LoginRequestType, { rejectWithValue }) => {
    try {
      const { data } = await postLogin(loginBody);
      // store user's token in local storage
      localStorage.setItem('userToken', data.token.accessToken);
      return data;
    } catch (error: any) {
      // return custom error message from API if any
      if (error.response && error.response.data.message) {
        return rejectWithValue(error.response.data.message);
      }
      return rejectWithValue(error.message);
    }
  },
);

// update
export const postUserData = createAsyncThunk<
null,
UserInfoType
>(
  'auth/update',
  async (newUserData: UserInfoType, { rejectWithValue }) => {
    try {
      const { data } = await (loginBody);
      // store user's token in local storage
      localStorage.setItem('userToken', data.token.accessToken);
      return data;
    } catch (error: any) {
      // return custom error message from API if any
      if (error.response && error.response.data.message) {
        return rejectWithValue(error.response.data.message);
      }
      return rejectWithValue(error.message);
    }
  },
);
