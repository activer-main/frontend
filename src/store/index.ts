import { configureStore } from '@reduxjs/toolkit';
import { useDispatch, TypedUseSelectorHook, useSelector } from 'react-redux';
import { authApi } from './auth/authService';
import authReducer from './auth/authSlice';
import { activityApi } from './activity/activityService';

const store = configureStore({
  reducer: {
    auth: authReducer,
    [authApi.reducerPath]: authApi.reducer,
    [activityApi.reducerPath]: activityApi.reducer,
  },
  middleware: (getDefaultMiddleware) => getDefaultMiddleware()
    .concat(authApi.middleware)
    .concat(activityApi.middleware),
});

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof store.getState>;
// Inferred type: {posts: PostsState, comments: CommentsState, users: UsersState}
export type AppDispatch = typeof store.dispatch;

// getting dispatch and selector type
export const useAppDispatch: () => AppDispatch = useDispatch;
export const useAppSelector: TypedUseSelectorHook<RootState> = useSelector;
export default store;
