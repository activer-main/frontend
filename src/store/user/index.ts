import {
  createSlice,
} from '@reduxjs/toolkit';
import type { RootState } from 'store';
import { UserDataType } from '../../types/user';
import { registerUser, userLogin } from './authAction';

const initialState: UserDataType = {
  loading: false,
  userInfo: null, // for user object
  userToken: null, // for storing the JWT
  error: null,
  success: false, // for monitoring the registration process.
};

const userSlice = createSlice({
  name: 'user',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      // regist
      .addCase(registerUser.pending, (state) => ({
        ...state,
        loading: true,
        error: null,
      }))
      .addCase(registerUser.fulfilled, (state) => ({
        ...state,
        loading: false,
        success: true,
      }))
      .addCase(registerUser.rejected, (state, { payload }) => ({
        ...state,
        loading: false,
        error: payload,
      }))
      // login
      .addCase(userLogin.pending, (state) => ({
        ...state,
        loading: true,
        error: null,
      }))
      .addCase(userLogin.fulfilled, (state, { payload }) => ({
        ...state,
        loading: false,
        userInfo: payload.user,
        userToken: payload.token,
      }))
      .addCase(userLogin.rejected, (state, { payload }) => ({
        ...state,
        loading: false,
        error: payload,
      }));
  },

});

export const getUserData = (state: RootState) => state.user;

export default userSlice.reducer;
