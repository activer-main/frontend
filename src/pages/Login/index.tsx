import React, { useEffect } from 'react';
import { useAppDispatch, useAppSelector } from 'store';
import { userLogin } from 'store/auth/authAction';
import { useNavigate } from 'react-router-dom';
import Button from 'components/Button';
import FormInput from 'components/FormInput';
import { LoginFormDataType } from 'types/user';
import './index.scss';
import { PASSWORD_PATTERN } from 'utils/pattern';

function Login() {
  const { loading, userInfo } = useAppSelector((state) => state.auth);
  const navigate = useNavigate();
  const dispatch = useAppDispatch();

  // redirect authenticated user to profile screen
  useEffect(() => {
    if (userInfo) {
      navigate(-1);
    }
  }, [navigate, userInfo]);

  const handleSubmit = (event: React.SyntheticEvent) => {
    event.preventDefault();
    const target = event.target as typeof event.target & LoginFormDataType;
    dispatch(userLogin({
      email: target.email.value,
      password: target.password.value,
    }));
  };

  return (
    <form onSubmit={handleSubmit} className="login">
      <h1 className="login__header">登入</h1>

      <FormInput
        className="login__account"
        name="email"
        label="帳號"
        type="email"
        title="請輸入正確帳號格式"
        id="email"
        required
      />

      <FormInput
        className="login__password"
        name="password"
        label="密碼"
        type="password"
        pattern={PASSWORD_PATTERN}
        title="密碼至少八位字元，需要包含至少一個數字、一個大寫英文、一個小寫英文、一個特殊字元: !@#$%"
        id="password"
        required
      />

      <div className="login__control">
        <Button
          type="submit"
          text={loading ? 'spining' : '登入'}
          className="button"
          disabled={loading}
        />
        <Button
          type="button"
          text="註冊"
          variant={{ outline: true }}
          className="button"
          disabled={loading}
          onClick={() => navigate('/register')}
        />

      </div>

    </form>
  );
}
export default Login;
