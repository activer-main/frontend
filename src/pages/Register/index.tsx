import * as React from 'react';
import Avatar from '@mui/material/Avatar';
import CssBaseline from '@mui/material/CssBaseline';
import TextField from '@mui/material/TextField';
import Grid from '@mui/material/Grid';
import Box from '@mui/material/Box';
import LockOutlinedIcon from '@mui/icons-material/LockOutlined';
import Typography from '@mui/material/Typography';
import LoadingButton from '@mui/lab/LoadingButton';
import Container from '@mui/material/Container';
import { useAppDispatch, useAppSelector } from 'store';
import { selectUserData } from 'store/auth/authSlice';
import { toast } from 'react-toastify';
import { registerUser } from 'store/auth/authAction';
import { useNavigate } from 'react-router-dom';
import BorderColorIcon from '@mui/icons-material/BorderColor';
import Link from '@mui/material/Link';
import {
  IconButton, InputAdornment,
} from '@mui/material';
import AccountCircle from '@mui/icons-material/AccountCircle';
import LockIcon from '@mui/icons-material/Lock';
import Visibility from '@mui/icons-material/Visibility';
import VisibilityOff from '@mui/icons-material/VisibilityOff';
import LockClockIcon from '@mui/icons-material/LockClock';
import { USERNAME_PATTERN } from 'utils/pattern';
import _ from 'lodash';

export default function Register() {
  const { loading } = useAppSelector(selectUserData);
  const dispatch = useAppDispatch();
  const navigate = useNavigate();

  const [showPassword, setShowPassword] = React.useState(false);
  const [showConfirmPassword, setShowConfirmPassword] = React.useState(false);
  const [validateError, setValidateError] = React.useState<string[]>([]);

  const handleClickShowPassword = () => setShowPassword((show) => !show);
  const handleClickShowConfirmPassword = () => setShowConfirmPassword((show) => !show);

  const handleMouseDownPassword = (event: React.MouseEvent<HTMLButtonElement>) => {
    event.preventDefault();
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    const data = new FormData(event.currentTarget);
    if (
      data.get('password') !== data.get('confirmPassword')
    ) {
      toast.error('密碼不相同');
      return;
    }

    dispatch(registerUser({
      username: data.get('username') as string,
      email: data.get('email') as string,
      password: data.get('password') as string,
    }))
      .unwrap()
      .then(() => {
        navigate('/user/profile');
      })
      .catch((error: any) => toast.error(error.message));
  };

  const handleValidate = (event:React.ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => {
    if (event.target.name === 'username') {
      if (event.target.value.match(USERNAME_PATTERN)) {
        setValidateError(_.filter(validateError, (item) => item !== 'username'));
      } else {
        setValidateError(_.uniq(_.concat(validateError, event.target.name)));
      }
    }
  };

  return (
    <Container component="main" maxWidth="xs">
      <CssBaseline />
      <Box
        sx={{
          marginTop: 8,
          display: 'flex',
          flexDirection: 'column',
          alignItems: 'center',
        }}
      >
        <Avatar sx={{ m: 1, bgcolor: 'secondary.main' }}>
          <LockOutlinedIcon />
        </Avatar>
        <Typography component="h1" variant="h4">
          註冊
        </Typography>
        <Box component="form" noValidate onSubmit={handleSubmit} sx={{ mt: 3 }}>
          <Grid container spacing={2}>
            <Grid item xs={12}>
              <TextField
                error={validateError.includes('username')}
                autoComplete="username"
                name="username"
                required
                fullWidth
                id="username"
                label="使用者名稱"
                helperText="使用者名稱須為2-16字中英數字"
                autoFocus
                inputProps={{
                  pattern: USERNAME_PATTERN,
                  startAdornment: (
                    <InputAdornment position="start">
                      <BorderColorIcon />
                    </InputAdornment>
                  ),
                }}
                onChange={(e) => handleValidate(e)}
              />
            </Grid>

            <Grid item xs={12}>
              <TextField
                required
                fullWidth
                id="email"
                label="帳號"
                name="email"
                autoComplete="email"
                InputProps={{
                  startAdornment: (
                    <InputAdornment position="start">
                      <AccountCircle />
                    </InputAdornment>
                  ),
                }}
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                required
                fullWidth
                name="password"
                label="密碼"
                type={showPassword ? 'text' : 'password'}
                id="password"
                autoComplete="new-password"
                InputProps={{
                  startAdornment: (
                    <InputAdornment position="start">
                      <LockIcon />
                    </InputAdornment>
                  ),
                  endAdornment: (
                    <InputAdornment position="end" sx={{ mr: 1 }}>
                      <IconButton
                        aria-label="toggle password visibility"
                        onClick={handleClickShowPassword}
                        onMouseDown={handleMouseDownPassword}
                        edge="end"
                      >
                        {showPassword ? <VisibilityOff /> : <Visibility />}
                      </IconButton>
                    </InputAdornment>
                  ),
                }}
              />
            </Grid>
            <Grid item xs={12}>
              <TextField
                required
                fullWidth
                name="confirmPassword"
                type={showConfirmPassword ? 'text' : 'password'}
                label="確認密碼"
                id="confirmPassword"
                autoComplete="new-password"
                InputProps={{
                  startAdornment: (
                    <InputAdornment position="start">
                      <LockClockIcon />
                    </InputAdornment>
                  ),
                  endAdornment: (
                    <InputAdornment position="end" sx={{ mr: 1 }}>
                      <IconButton
                        aria-label="toggle password visibility"
                        onClick={handleClickShowConfirmPassword}
                        onMouseDown={handleMouseDownPassword}
                        edge="end"
                      >
                        {showConfirmPassword ? <VisibilityOff /> : <Visibility />}
                      </IconButton>
                    </InputAdornment>
                  ),
                }}
              />
            </Grid>

          </Grid>
          <LoadingButton
            loading={loading}
            loadingPosition="end"
            type="submit"
            fullWidth
            variant="contained"
            sx={{ mt: 3, mb: 2 }}
          >
            註冊
          </LoadingButton>
          <Grid container justifyContent="flex-end" alignItems="flex-end" spacing={1}>
            <Grid item>
              已有帳號?
            </Grid>
            <Grid item>
              <Link
                component="button"
                underline="none"
                sx={{
                  color: 'secondary.main',
                  '&:hover': {
                    color: 'secondary.dark',
                  },
                }}
                onClick={() => navigate('/login')}
              >
                立即登入
              </Link>
            </Grid>
          </Grid>
        </Box>
      </Box>
    </Container>
  );
}
