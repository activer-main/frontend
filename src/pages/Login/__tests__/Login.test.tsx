import React from 'react';
import { fireEvent, screen, waitFor } from '@testing-library/react';
import { renderWithProviders } from 'utils/testUtils';
import { setupStore } from 'store';
import { userLogin } from 'store/auth/authAction';
import { createServer } from 'test/server';
import {
  MemoryRouter, Routes, Route,
} from 'react-router-dom';
import { mockUserData } from 'test/data/user';
import userEvent from '@testing-library/user-event';
import ReactTestRenderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import { act } from 'react-dom/test-utils';

import Register from 'pages/Register';
import Login from '../index';

// Mock Server
createServer([
  {
    path: 'http://220.132.244.41:5044/api/User/auth/signin',
    method: 'post',
    res: () => mockUserData,
  },
]);

// Test
describe('Login component', () => {
  it('Login Page render correctly', () => {
    // render component
    renderWithProviders(
      <MemoryRouter initialEntries={['/login']}>
        <Login />
      </MemoryRouter>,
    );

    // assertion
    expect(screen.getByRole('heading', { name: '登入' })).toBeInTheDocument();
    expect(screen.getByRole('textbox', { name: '帳號' })).toBeInTheDocument();
    expect(screen.getByLabelText('密碼')).toBeInTheDocument();
    expect(screen.getByRole('button', { name: '登入' })).toBeInTheDocument();
    expect(screen.getByRole('button', { name: '註冊' })).toBeInTheDocument();
  });

  it('should match snapshot', () => {
    // render component
    const store = setupStore();
    const component = (
      <Provider store={store}>
        <MemoryRouter initialEntries={['/login']}>
          <Routes>
            <Route path="login" element={<Login />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );

    const renderer = ReactTestRenderer.create(component);

    // assertion
    expect(renderer.toJSON()).toMatchSnapshot();
  });

  it('should render to previous page if user is already authenticated', async () => {
    // setup login body
    const email = 'test@test.com';
    const password = 'Test1234!';

    // setup redux store
    const store = setupStore();
    await store.dispatch(userLogin({ email, password }));

    // render component
    renderWithProviders(
      <MemoryRouter initialEntries={['/login', '/test']}>
        <Routes>
          <Route path="/login" element={<Login />} />
          <Route path="/test" element={<h1>test</h1>} />
        </Routes>
      </MemoryRouter>,
      { store },
    );

    // assertion not in login page
    expect(screen.queryByRole('heading', { name: '登入' })).not.toBeInTheDocument();
  });

  it('should render to register page', async () => {
    // render component
    renderWithProviders(
      <MemoryRouter initialEntries={['/login', '/register']}>
        <Routes>
          <Route path="/login" element={<Login />} />
          <Route path="/register" element={<Register />} />
        </Routes>
      </MemoryRouter>,
    );

    // query register button
    const registerButton = screen.getByRole('button', { name: '註冊' });

    // click register button
    userEvent.click(registerButton);

    // assertion
    expect(screen.getByRole('heading', { name: '註冊' })).toBeInTheDocument();
  });

  it('Should submit data correctly', async () => {
    // render component
    const { store, getByLabelText } = renderWithProviders(
      <MemoryRouter initialEntries={['/login']}>
        <Login />
      </MemoryRouter>,
    );

    const emailInput = screen.getByLabelText('帳號');
    const passwordInput = getByLabelText('密碼');
    const submitButton = screen.getByRole('button', { name: '登入' });

    act(() => {
      fireEvent.change(emailInput, { target: { value: mockUserData.user.email } });
      fireEvent.change(passwordInput, { target: { value: 'Password123!' } });
    });

    fireEvent.click(submitButton);

    expect(store.getState().auth.loading).toBe(true);
    expect(store.getState().auth.userInfo).toBe(null);

    // Wait for the API call to complete and update the store
    waitFor(() => {
      console.log(store.getState().auth);
      expect(store.getState().auth.loading).toBe(false);
    });
  });
});
