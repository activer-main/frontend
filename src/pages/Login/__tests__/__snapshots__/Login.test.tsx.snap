// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should match snapshot 1`] = `
<form
  className="login"
  onSubmit={[Function]}
>
  <h1
    className="login__header"
  >
    登入
  </h1>
  <div
    className="form-input login__account"
  >
    <label
      className="form-input__label"
      htmlFor="form-input-email"
    >
      帳號
    </label>
    <input
      className="form-input__input"
      id="form-input-email"
      name="email"
      onChange={[Function]}
      required={true}
      type="email"
      value=""
    />
    <span
      className="form-input__error-message"
    />
  </div>
  <div
    className="form-input login__password"
  >
    <label
      className="form-input__label"
      htmlFor="form-input-password"
    >
      密碼
    </label>
    <input
      className="form-input__input"
      id="form-input-password"
      name="password"
      onChange={[Function]}
      pattern="^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#$%]).{8,24}$"
      required={true}
      type="password"
      value=""
    />
    <span
      className="form-input__error-message"
    />
  </div>
  <div
    className="login__control"
  >
    <button
      className="button button--primary button"
      disabled={false}
      type="submit"
    >
      登入
    </button>
    <button
      className="button button--primary button--outline button"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      註冊
    </button>
  </div>
</form>
`;
